# PRS_cal_pipeline
Calculate PRS by plink, now only try run one score not multiple score in one weight file

Extract SNPs from whole variants to save time, especially when need calcuate multiple scores

The format of weight score:
SNP(CHR:POS:A1:A2 (A1 and A2 are alphabetically ordered)), Chr, Pos, effect_allele, other_allele, effect_weight
and example file can be found in example/example.weight.txt

1. 01.extract.bgen.last.sh / 01.extract.bgen.first.sh

   Extract the bed region from the orignial imputation or whole genome wide genotype file to save time

   For MGB I should choose the 01.extract.bgen.last.sh, but for UKB I should choose 01.extract.bgen.first.sh


2. 02.setID.sh

   Set the variant ID into chr@:#:$r:$a

3. 03.updateID.sh

   Update the variant ID into this format(CHR:POS:A1:A2 (A1 and A2 are alphabetically ordered)) to match the weight

4. 04.calculate.score.sh

   Calculate score by
   --score $5 1 4 header-read list-variants ignore-dup-ids cols='sid,nallele,dosagesum,scoresums' \
   --score-col-nums $6

   Here I use scoresums not the defult average (https://groups.google.com/g/prsice/c/hy-C66uo8ok?pli=1), AVG has the benefit of accounting for 
   missingness though in most case, with proper QC, the AVG and SUM score should be more or less the same, as such the result should not change.

5. 05.combinechr.R

   combine the prs generated by each chr into all chrmosome

Also, for UKB and MGB example run whole code can be found in example/UKB.score.sh and example/MGB.score.sh

