# PRS_cal_pipeline
## Calculate PRS by plink, now only try run one score not multiple score in one weight file

Extract SNPs from whole variants to save time, especially when need calcuate multiple scores

The format of weight score:
SNP(CHR:POS:A1:A2 (***A1 and A2 are alphabetically ordered***)), Chr, Pos, effect_allele, other_allele, effect_weight
and example file can be found in example/example.weight.txt

1. [01.extract.bgen.last.sh](https://github.com/suiyangsun/PRS_cal_pipeline/blob/main/01.extract.bgen.last.sh) / [01.extract.bgen.first.sh](https://github.com/suiyangsun/PRS_cal_pipeline/blob/main/01.extract.bgen.first.sh)

   Extract the bed region from the orignial imputation or whole genome wide genotype file to save time

   For MGB I should choose the 01.extract.bgen.last.sh, but for UKB I should choose 01.extract.bgen.first.sh


2. [02.setID.sh](https://github.com/suiyangsun/PRS_cal_pipeline/blob/main/02.setID.sh)

   Set the variant ID into chr@:#:$r:$a

3. [03.updateID.sh](https://github.com/suiyangsun/PRS_cal_pipeline/blob/main/03.updateID.sh)

   Update the variant ID into this format(CHR:POS:A1:A2 (***A1 and A2 are alphabetically ordered***)) to match the weight

4. [04.calculate.score.sh](https://github.com/suiyangsun/PRS_cal_pipeline/blob/main/04.calculate.score.sh)

   Calculate score by
   --score $5 1 4 header-read list-variants ignore-dup-ids cols='sid,nallele,dosagesum,scoresums' \
   --score-col-nums $6

   Here I use ***scoresums*** not the defult average (https://groups.google.com/g/prsice/c/hy-C66uo8ok?pli=1), because when I calculate by chr, it's not very fair using AVG. AVG has the benefit of accounting for 
   missingness though in most case, with proper QC, the AVG and SUM score should be more or less the same, as such the result should not change.

6. [05.combinechr.R](https://github.com/suiyangsun/PRS_cal_pipeline/blob/main/05.combinechr.R)

   combine the prs generated by each chr into all chrmosome

Also, for UKB and MGB example run whole code can be found in example/UKB.score.sh and example/MGB.score.sh

## Process the raw PRS to get performance, like OR, AUC

[Residuals_YS.R](https://github.com/suiyangsun/PRS_cal_pipeline/blob/main/Residuals_YS.R)

Extract residulas from linear regression. Like PRS ~ PC1 + PC2

[Scale.R](https://github.com/suiyangsun/PRS_cal_pipeline/blob/main/Scale.R)

Normolize column by scale.

[KeyMapReplacer.py](https://github.com/suiyangsun/PRS_cal_pipeline/blob/main/KeyMapReplacer.py) 

Merge two files, like zcat score/{1}-sscore.gz |python KeyMapReplacer.py -k1 -a NA -p<(cat $pheno) -x

[wcut.py](https://github.com/suiyangsun/PRS_cal_pipeline/blob/main/wcut.py)

Get the columns either by their name or by column number, like -f4,3; -f4,2-1, -f5-2; -t title1,title2,title3

[GlmRegression_YS.R](https://github.com/suiyangsun/PRS_cal_pipeline/blob/main/GlmRegression_YS.R)

Get the performance of PRS, including AUC, Coefficients, Incremental_Model_Rsq, Pearson_Correlation

**Notice: Regeression link gaussian for linear regression (continuous phenotype), bionomial for logistic regression (binary phenotype).**

The example is:
```
name="BMI"
#Regeression link gaussian for linear regression (continuous phenotype), bionomial for logistic regression (binary phenotype).
link="gaussian"

full_model="$name~adjNormPRS+age+inferred_gender+genotyping_array+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10"
null_model="$name~age+inferred_gender+genotyping_array+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10"

zcat score/sscore.gz | \
  ##change the #IID to IID
  sed 's/#IID/IID/g' | \
  ##merge the score with phenotype
  python KeyMapReplacer.py -k1 -a NA -p<(cat $pheno) -x | \
  ##change the raw PRS name into PRS
  sed 's/effect_weight_SUM/PRS/g' | \
  ##only get the PCs to minimum to remove NA
  wcut -t'IID,PRS,PC1,PC2,PC3,PC4,PC5,PC6,PC7,PC8,PC9,PC10' | \
  ##regress out top 10 PCs or 4 PCs
  Rscript Residuals_YS.R -f 'PRS~PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10' -t adjPRS | \
  ##normalize adjPRS
  Rscript Scale.R -c adjPRS -t adjNormPRS | \
  ##get the IID and PRSs
  python wcut.py -t 'IID,PRS,adjPRS,adjNormPRS' | \
  ##merge with phenotype
  python KeyMapReplacer.py -k1 -a NA -p<(cat $pheno) -x | \
  ##only get the phenotype needed for fit the model
  python wcut.py -t '$name,adjNormPRS,age,inferred_gender,genotyping_array,PC1,PC2,PC3,PC4,PC5,PC6,PC7,PC8,PC9,PC10' | \
  Rscript GlmRegression_YS.R -f $full_model -m $link -n $null_model -r y -p y -a AUC -t $name |
  tee result/$name.log
```









